omnigan/trainer.py

```diff
@@ -1773,3 +1292,280 @@ def eval_images(self, mode, domain):
+                loss = self.losses["G"]["tasks"]["m"]["gi"](pred_prob, target)
+                loss *= self.opts.train.lambdas.G["m"]["gi"]
+                full_loss += loss
+                self.logger.losses.gen.task["m"]["gi"]["r"] = loss.item()
+
+            # Painter loss
+            if self.use_pm4l and for_ == "G":
+                pl4m_loss = self.painter_loss_for_masker(x, pred_prob)
+                pl4m_loss *= self.opts.train.lambdas.G.m.pl4m
+                full_loss += pl4m_loss
+                self.logger.losses.gen.task.m.pl4m.r = pl4m_loss.item()
+
+            if self.opts.gen.m.use_minent:
+                # MinEnt loss
+                loss = self.losses["G"]["tasks"]["m"]["minent"](prob.to(self.device))
+                loss *= self.opts.train.lambdas.advent.ent_main
+                full_loss += loss
+                self.logger.losses.gen.task["m"]["minent"]["r"] = loss.item()
+
+        if self.opts.gen.m.use_advent:
+            # AdvEnt loss
+            if for_ == "D":
+                domain_label = domain
+                logger = {}
+                loss_func = self.losses["D"]["advent"]
+                prob = prob.detach()
+                weight = self.opts.train.lambdas.advent.adv_main
+            else:
+                domain_label = "s"
+                logger = self.logger.losses.gen.task["m"]["advent"]
+                loss_func = self.losses["G"]["tasks"]["m"]["advent"]
+                weight = self.opts.train.lambdas.advent.adv_main
+
+            if for_ == "D" or domain == "r":
+                loss = loss_func(
+                    prob.to(self.device),
+                    self.domain_labels[domain_label],
+                    self.D["m"]["Advent"],
+                )
+                loss *= weight
+                full_loss += loss
+                logger[domain] = loss.item()
+
+            if for_ == "D":
+                if self.opts.dis.s.gan_type == "GAN" or "WGAN_norm":
+                    pass
+                elif self.opts.dis.s.gan_type == "WGAN":
+                    for p in self.D["s"]["Advent"].parameters():
+                        p.data.clamp_(
+                            self.opts.dis.s.wgan_clamp_lower,
+                            self.opts.dis.s.wgan_clamp_upper,
+                        )
+                elif self.opts.dis.s.gan_type == "WGAN_gp":
+                    prob_need_grad = autograd.Variable(prob, requires_grad=True)
+                    d_out = self.D["s"]["Advent"](prob_need_grad)
+                    gp = get_WGAN_gradient(prob_need_grad, d_out)
+                    gp_loss = self.opts.train.lambdas.advent.WGAN_gp * gp
+                    full_loss += gp_loss
+                else:
+                    raise NotImplementedError
+
+        return full_loss
+
+    def painter_loss_for_masker(self, x, m):
+        # pl4m loss
+        # painter should not be updated
+        for param in self.G.painter.parameters():
+            param.requires_grad = False
```

Comment:
i think this looks good to me

Reply:
I agree that it's good! `detach()` creates a new tensor while `requires_grad()` just tells the model not to compute the gradient for this parameter, so I think it's the best way to do it.

Reply:
Cool